### PR TITLE
Only one development environment

### PR DESCRIPTION
**Pull request by garbas**

### CODE REVIEW

File: shells/default/default.nix [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    ];
+in
+  mkShell (
+    {
+      name = "flox-dev";
+
+      inputsFrom = [
+        flox-pkgdb
+        flox-env-builder
+        flox
```
Comment (ysndr) [inline shells/default/default.nix:64]: doesn't `flox` have `flox-pkgdb` and `flox-env-builder` as dependencies?
should this rather be something along this?

```
[
	flox-pkgdb
    flox-env-builder
    (flox.override {
		flox-pkgdb = null;
		flox-env-builder = null;
	})
]
```

File: checks/pre-commit-check/default.nix [hunk 2]
```diff
@@ -1,23 +1,12 @@
...
-    };
-  })
-  // {passthru = {inherit rustfmt;};}
+  shellHooks,
+}:
+shellHooks.lib.${system}.run {
+  src = builtins.path {path = ../..;};
+  hooks = {
+    alejandra.enable = true;
+    rustfmt.enable = true;
```
[the remark this answers is not in the document]
Comment (garbas) [reply checks/pre-commit-check/default.nix:9]: in the overlays (see `flake.nix`).

If we do it here we need to pass it every time around.

File: flake.nix [hunk 3]
```diff
@@ -121,27 +121,20 @@
...
       flox-tests-end2end = final.flox-tests.override {
         name = "flox-tests-end2end";
         testsDir = "/tests/end2end";
       };
-      flox-tests-end2end-dev = final.flox-tests.override {
-        name = "flox-tests-end2end";
-        testsDir = "/tests/end2end";
-        FLOX_CLI = null;
+
+      rustfmt = prev.rustfmt.override {asNightly = true;};
```
[the remark this answers is not in the document]
Comment (garbas) [reply flake.nix:135]: You found it :)

File: shells/default/default.nix (hunk 1 above)
Comment (garbas) [reply shells/default/default.nix:68]: Yes, but in this case (development environment) we want to use development versions of those projects. Setting things to null will use things from path.

In CI we actually need the opposite.

File: Justfile [hunk 4]
```diff
@@ -1,19 +1,18 @@
...
-bats_invocation := "nix --extra-experimental-features nix-command --extra-experimental-features flakes run '.#flox-tests-dev' -- --flox ${PWD}/target/debug/flox"
 
 _default:
     @just --list --unsorted
 
 # Run the 'bats' test suite
 bats-tests +bats_args="":
     @cargo build -q
-    @{{bats_invocation}} -- {{bats_args}}
+    @flox-tests
```
[the remark this answers is not in the document]
Comment (garbas) [reply Justfile:10]: With Alex's work (see https://github.com/flox/flox/pull/566) this Justfile will/should go away.

File: shells/default/default.nix (hunk 1 above)
Comment (ysndr) [reply shells/default/default.nix:68]: It's just, you dont want to rebuild `flox-pkgdb` when pkgdb files changed and you ope the dev shell?

In CI i get you want to use the built ones for testing `cargo build flox`

File: pkgdb/Makefile [hunk 5]
```diff
@@ -633,13 +633,8 @@ cc-check: $(TESTS:.cc=)
...
 
 #: Run all bats tests
-BATS_FILE ?= $(MAKEFILE_DIR)/tests
 bats-check: bin $(TEST_UTILS)
-	PKGDB="$(MAKEFILE_DIR)/bin/pkgdb"                        \
-	IS_SQLITE3="$(MAKEFILE_DIR)/tests/is_sqlite3"            \
-	SEARCH_PARAMS="$(MAKEFILE_DIR)/tests/search-params"      \
-	  bats --print-output-on-failure --verbose-run --timing  \
-	       "$(BATS_FILE)";
+	flox-pkgdb-tests
```
Comment (aakropotkin) [inline pkgdb/Makefile:637]: non-blocking: this is fine as long as the underlying test script does not depend on other _built_ packages from `github:flox/flox`, and it respects environment variables passed in as `FOO=1 flox-pkgdb-tests`.

File: pkgdb/tests/ga-registry.bats [hunk 6]
```diff
@@ -176,7 +176,7 @@ setup_file() {
 # bats test_tags=manifest:ga-registry, lock:ga-registry, manifest:global
 
 @test "'pkgdb manifest lock --ga-registry' rejects global manifest registry" {
-  run $PKGDB manifest lock --ga-registry                                     \
+  run $PKGDB_BIN manifest lock --ga-registry                                     \
                            --global-manifest "$TDATA/global-manifest0.toml"  \
```
Comment (aakropotkin) [inline pkgdb/tests/ga-registry.bats:180]:
```suggestion
  run $PKGDB_BIN manifest lock --ga-registry                                 \
                           --global-manifest "$TDATA/global-manifest0.toml"  \
```

File: pkgdb/tests/search.bats [hunk 7]
```diff
@@ -13,8 +13,8 @@ load setup_suite.bash;
 setup_file() {
   export TDATA="$TESTS_DIR/data/search";
 
-  export PKGDB_CACHEDIR="$BATS_FILE_TMPDIR/pkgdbs";
-  echo "PKGDB_CACHEDIR: $PKGDB_CACHEDIR" >&3;
+  export PKGDB_BIN_CACHEDIR="$BATS_FILE_TMPDIR/pkgdbs";
+  echo "PKGDB_BIN_CACHEDIR: $PKGDB_BIN_CACHEDIR" >&3;
```
Comment (aakropotkin) [inline pkgdb/tests/search.bats:17]:
```suggestion
  export PKGDB_CACHEDIR="$BATS_FILE_TMPDIR/pkgdbs";
  echo "PKGDB_CACHEDIR: $PKGDB_CACHEDIR" >&3;
```

File: pkgdb/tests/sqlite3.bats [hunk 8]
```diff
@@ -25,48 +25,48 @@ setup_file() {
...
+  if [[ -z "${PKGDB_IS_SQLITE3_BIN:-}" ]]; then
     repo_root_setup;
-    IS_SQLITE3="$TESTS_DIR/is_sqlite3";
-    if ! [[ -x "$IS_SQLITE3" ]]; then
+    PKGDB_IS_SQLITE3_BIN="$TESTS_DIR/is_sqlite3";
+    if ! [[ -x "$PKGDB_IS_SQLITE3_BIN" ]]; then
       (
         cd "${REPO_ROOT?}" >/dev/null 2>&1||exit 1;
-        nix develop -c make -j8 tests/is_sqlite3;
+        make -j8 tests/is_sqlite3;
```
Comment (aakropotkin) [inline pkgdb/tests/sqlite3.bats:34]:
```suggestion
        make -j tests/is_sqlite3;
```

File: pkgs/flox-env-builder-tests/default.nix [hunk 9]
```diff
@@ -28,50 +31,89 @@
...
 in
-  writeShellScriptBin "flox-env-builder-tests" ''
-    set -x -euo pipefail
+  # TODO: we should run tests against different shells
+  writeShellScriptBin PROJECT_NAME ''
+    set -euo pipefail
 
-    export PATH="${lib.makeBinPath paths}"
+    # Find root of the subproject if not specified
+    PROJECT_TESTS_DIR=${PROJECT_TESTS_DIR}
```
Comment (aakropotkin) [inline pkgs/flox-env-builder-tests/default.nix:39]:
```suggestion
    PROJECT_TESTS_DIR="${PROJECT_TESTS_DIR}";
```

File: pkgs/flox-env-builder-tests/default.nix [hunk 10]
```diff
@@ -28,50 +31,89 @@
...
+    export PROJECT_TESTS_DIR
+
+    # TODO: we shouldn't do this but rather use absolute paths
+    # Look if we can use https://github.com/abathur/resholve
+    export PATH="$PROJECT_PATH${lib.makeBinPath paths}"
+
+    # Copy PROJECT_TESTS_DIR to temporary directory
+    WORKDIR=$(mktemp -d -t ${PROJECT_NAME}-XXXXXX)
+    cp -R $PROJECT_TESTS_DIR/* $WORKDIR
     cd $WORKDIR
```
Comment (aakropotkin) [inline pkgs/flox-env-builder-tests/default.nix:63]:
```suggestion
    WORKDIR="$( mktemp -d -t ${PROJECT_NAME}-XXXXXX; )";
    cp -R "$PROJECT_TESTS_DIR/"* "$WORKDIR";
    cd "$WORKDIR";
```

File: tests/bads.bats
Comment (aakropotkin) [inline tests/bads.bats:1]: TODO: Delete this file


This is fine to do in a later PR.

**Review by aakropotkin — CHANGES_REQUESTED** [review]

**Review by aakropotkin — CHANGES_REQUESTED** [review]

**Comment by aakropotkin** [comment]
LGTM after the marked changes

File: tests/bads.bats
Comment (garbas) [reply tests/bads.bats:1]: done

**Review by aakropotkin — APPROVED** [review]
Test failure looked like a typo, feel free to merge once that's passing.

**Pull request merged**